TMEDIA-852 Image component

JackHowa opened the pull request:
## Ticket

- [TMEDIA-852](https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-843&assignee=5e387d6a1b1d910e5dfd7a9b)

## Description

Since the image component does not currently handle responsive images, we’re going to utilize the power of v2 resizer to do so. 

Please remember that, like discussed in tech party, network conditions, browser, and screen web pixel density may affect which image is displayed.  

We want to make this forward-compatible with any resizer v2 improvements or features. So this will accept fields via an object that can theoretically pass in any field.
## Acceptance Criteria

- [x] Remove width and height fields from existing Image component in the arc themes components repo
- [x] Add resizerOptions field that takes in an object with keys for current or future resizer v2 query parameters.
- [x] If empty resizerOptions, then render raw image with no width or height.
- [x] The width and height properties of the img tag are determined by what’s inside the resizerOptions
- [x] Create storybook examples for no width, no height, width and height, and various available combinations of resizer options
- [x] Add a responsiveImages array of numbers. These are the width descriptor. Must be positive integers
- [x] The responsiveImages and the resizerOptions will be responsible for creating the srcSet attribute in the image tag 
- [x] The width in the resizerOptions would multiply by each responsiveImages integer to create a srcSet's url with the filters as well as the width descriptor 
- [x] Cleanup some of the logic around generating a srcset. 
- [x] Ensure that, if width is not required in the resizedOptions, then the image responsive array will be solely responsible for setting the width. 
- [x] Don't render the srcset if empty responsive images array.
- [ ] This srcSet array mapper should be in the utils folder for reusability -> investigating patterns now, but open to feedback
- [x] Take in a sizes field that is an array of objects 
- [x] Each item needs a mediaCondition and a sourceSizeValue, except for one isDefault
- [x] There must be one isDefault object. This will be the last in the string of the generated sizes property 
- [x] If there is more than one isDefault, then use the first one specified


## Test Steps

1. Checkout branch - `git checkout TMEDIA-852-image-component`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Go to the image component stories

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

JackHowa committed:
feat: Ensure intrinsic sizes and srcset respected while still using width and height for cls

JackHowa committed:
feat: Take in resizer options of height and width

Remove width and height fields from existing Image component in the arc themes components repo. The width and height properties of the img tag are determined by what’s inside the resizerOptions

JackHowa committed:
feat: Add resizer options whatever passed in

JackHowa committed:
docs: Add query parameters docs for available features

JackHowa committed:
feat: Add resizer url to prefix the image source

JackHowa committed:
feat: Handle responsive images preliminary without height and width empty fallbacks

JackHowa committed:
feat: Without an auth token, render the raw image. Add auth token logic and resizing for sizes

JackHowa committed:
style: Update styles and linting for the tabs

JackHowa committed:
feat: Add aspect ratio zero check. Add responsive images check

JackHowa committed:
test: Update to check for responsive images positive integers

JackHowa committed:
docs: Add storybook of an image with an auth token and resizer

JackHowa committed:
refactor: Use dynamic key for height with aspect ratio. Group related code together

JackHowa committed:
style: Remove indent conflicting rule using eslint-config-prettier cli helper tool

JackHowa committed:
chore: Remove other conflicting eslint rules with prettier

JackHowa committed:
feat: Add sizes array of objects for responsive images

JackHowa committed:
fix: Use the first isDefault object size provided

JackHowa committed:
feat: Add responsive images. Handle sizes options

JackHowa committed:
feat: Add viewport size units

nschubach commented on src/components/image/index.jsx:
```diff
@@ -2,35 +2,161 @@ import PropTypes from "prop-types";
+	}
+
+	// add all options except height and width
+	const stringOptionsWithoutHeightWidth = Object.keys(resizedOptions).reduce((acc, key) => {
+		if (key === "height" || key === "width") {
+			return acc;
+		}
+
+		if (acc === "") {
+			return `?${key}=${resizedOptions[key]}`;
+		}
+
+		return `${acc}&${key}=${resizedOptions[key]}`;
+	}, "");
```
If you remove the width and height from `resizedOptions` you can do:
`new URLSearchParams(resizedOptions).toString();`

If we can't remove width/height from resizedOptions, you could destructure them out:
```
const { width, height, ...resizeParams } = resizedOptions;
const urlSearchParams = new URLSearchParams(resizeParams).toString(); // ex: 'filter=filterHere&auth=abc123'

```

JackHowa replied:
will look into this -- thanks

JackHowa replied:
updated thanks  🆒

JackHowa replied:
let me know if this is what you imagined. I liked the width split out change

JackHowa replied:
all good I presume with approval?

nschubach commented on src/components/image/index.jsx:
```diff
@@ -2,35 +2,161 @@ import PropTypes from "prop-types";
+			src={defaultSrc}
+			width={width}
+			srcSet={
+				responsiveHeightsAndWidths.length > 0
+					? responsiveHeightsAndWidths
+							.map(
+								(responsiveImage) =>
+									`${srcWithOptionsWithoutHeightWidth}&width=${responsiveImage.width}${
+										responsiveImage?.height ? `&height=${responsiveImage.height}` : ""
+									} ${responsiveImage.width}w`
+							)
+							.join(", ")
+					: null
```
this can be written to not have to check the length:
```suggestion
				responsiveHeightsAndWidths
						.map(
							(responsiveImage) =>
								`${objectToQueryParams(responsiveImage)} ${responsiveImage.width}w`
						)
						.join(", ") || null
```
* not needing to check the length since a 0 length is a falsy string
* not needing the `srcWithOptionsWithoutHeightWidth` since removeFalsy removes falsy url variables ... may want to check for null or undefined instead since false, 0, or blank COULD be valid

With support from:
```
function objectToQueryParams(obj) {
	new URLSearchParams(removeFalsy(obj)).toString();
}
function removeFalsy(obj) {
	return Object.fromEntries(Object.entries(obj).filter(([_, v]) => v));
}
```

nschubach replied:
Curious though, is there a default for `width`?

JackHowa replied:
default for responsive images array is empty arr. for responsiveImage.width, the user has total control over that?

JackHowa replied:
updated

JackHowa replied:
I updated to remove the length check and use short-circuiting based on the feedback. I'm not sure I understand the objectToQueryParams function: we need the resizer, the image name itself, and the search params. 

The current `srcWithOptionsWithoutHeightWidth` has this now. I'm using search params method to construct that string already.

nschubach replied:
Doesn't really work with what you have now, but I have some other comments.  Resolving this one.

JackHowa committed:
refactor: Update based on nick's suggestion for utilizing search params method

via https://github.com/WPMedia/arc-themes-components/pull/112#discussion_r925968149

JackHowa committed:
refactor: Update docs based on Nick's comment on refactor for width and height

Via https://github.com/WPMedia/arc-themes-components/pull/112#discussion_r925968149

JackHowa committed:
refactor: Update to use short circuiting without using a length check, checking the output, rather than checking that there's anything to input

Via https://github.com/WPMedia/arc-themes-components/pull/112#discussion_r926015589

JackHowa commented on src/components/image/index.jsx:
```diff
@@ -2,20 +2,124 @@ import PropTypes from "prop-types";
+			src={defaultSrc}
+			width={width}
+			srcSet={
+				responsiveHeightsAndWidths
```
conflicting prettier and eslint tab controls. and I found this eslint-prettier-config cli tool to resolve the diffs

JackHowa commented on .eslintrc.js:
```diff
@@ -21,15 +21,12 @@ module.exports = {
 	rules: {
 		"global-require": "off",
-		indent: ["error", "tab"],
 		"no-tabs": [
```
had conflicting prettier and eslint rules: 

before 

conflicting eslint and prettier 

howardj@arc-20750 arc-themes-components % npx eslint-config-prettier .eslintrc.js 
The following rules are unnecessary or might conflict with Prettier:

- indent
- react/jsx-indent
- react/jsx-indent-props
howardj@arc-20750 arc-themes-components % 

after 

howardj@arc-20750 arc-themes-components % npx eslint-config-prettier .eslintrc.js
No rules that are unnecessary or conflict with Prettier were found.

JackHowa committed:
refactor: Extract inline prop logic out the jsx for readability

Based on Matthew's feedback

JackHowa committed:
docs: Update the title to be more clear on sizes behavior

JackHowa committed:
docs: Add more information on image use best practices

JackHowa commented on .eslintrc.js:
```diff
@@ -21,15 +21,12 @@ module.exports = {
 	plugins: ["react", "jest", "jsx-a11y", "react-hooks", "sort-exports"],
 	rules: {
 		"global-require": "off",
-		indent: ["error", "tab"],
```
todo: update this for the blocks as well in a separate pr

JackHowa replied:
no issue in the blocks around this

JackHowa replied:
howardj@arc-20750 fusion-theme-blocks % npx eslint-config-prettier .eslintrc.js
No rules that are unnecessary or conflict with Prettier were found.

JackHowa commented on .eslintrc.js:
```diff
@@ -21,15 +21,12 @@ module.exports = {
 				allowIndentationTabs: true,
 			},
 		],
-		"react/jsx-indent": [2, "tab"],
```
this was done because there were conflicts with prettier and eslint: 

before 

conflicting eslint and prettier 

howardj@arc-20750 arc-themes-components % npx eslint-config-prettier .eslintrc.js 
The following rules are unnecessary or might conflict with Prettier:

- indent
- react/jsx-indent
- react/jsx-indent-props
howardj@arc-20750 arc-themes-components % 

after 

howardj@arc-20750 arc-themes-components % npx eslint-config-prettier .eslintrc.js
No rules that are unnecessary or conflict with Prettier were found.

JackHowa committed:
feat: Update explanation of src and resizer slashes

matthewroach reviewed (approved):
Looks good - be nice to finally get to use v2 resizer in blocks

nschubach committed:
Simplify some of the loops for processing image data (#120)

JackHowa commented on src/components/image/index.jsx:
```diff
@@ -2,20 +2,102 @@ import PropTypes from "prop-types";
+	}
+
+	// "https://resizer.com/" + "image.jpg" + "?auth=secret&filter=true"
+	const srcWithOptionsWithoutHeightWidth = resizerURL.concat(
```
I suppose we're not sure that height and width are not on this resized options

JackHowa replied:
but it's not in the docs
